File: engine/flphys.c
#include "flphys.h"
#include "types.h"

#include <math.h>
#include <stdlib.h>
#include <string.h>

const char *phys_strerror(int result) {
    switch (result) {
    case PHYS_RES_OK:
        return "Success";
    case PHYS_RES_ERR_NULL_PTR:
        return "Null pointer error";
    case PHYS_RES_ERR_ZERO_DIST:
        return "Zero distance error";
    case PHYS_RES_ERR_ZERO_MASS:
        return "Zero mass error";
    default:
        return "Unknown error code";
    }
}

void pvec_set_scs(struct pvec *pvec, double len, double angle) {
    pvec->x = len * cos(angle);
    pvec->y = len * sin(angle);
}

double pvec_get_x(const struct pvec *pvec) {
    return pvec->x;
}

double pvec_get_y(const struct pvec *pvec) {
    return pvec->y;
}

void pvec_set_x(struct pvec *pvec, double x) {
    pvec->x = x;
}

void pvec_set_y(struct pvec *pvec, double y) {
    pvec->y = y;
}

double pvec_get_len(const struct pvec *pvec) {
    return sqrt(pvec->x * pvec->x + pvec->y * pvec->y);
}

double pvec_get_angle(const struct pvec *pvec) {
    return atan2(pvec->y, pvec->x);
}

struct pvec *pobj_ref_pos(struct pobj *pobj) {
    return &pobj->pos;
}

struct pvec *pobj_ref_mov(struct pobj *pobj) {
    return &pobj->mov;
}

double pobj_get_mass(const struct pobj *pobj) {
    return pobj->mass;
}

void pobj_set_mass(struct pobj *pobj, double mass) {
    pobj->mass = mass;
}

double pobj_get_radius(const struct pobj *pobj) {
    return pobj->radius;
}

void pobj_set_radius(struct pobj *obj, double radius) {
    obj->radius = radius;
    obj->area = PHYS_PI * radius * radius;
    obj->volume = (4.0 / 3.0) * PHYS_PI * radius * radius * radius;
}

static int pobj_run(struct pobj *obj, double time) {
    if (obj->mass == 0) {
        return PHYS_RES_ERR_ZERO_MASS;
    }
    struct pvec acceleration = {
        .x = obj->force.x / obj->mass,
        .y = obj->force.y / obj->mass,
    };

    obj->pos.x += (obj->mov.x + acceleration.x * time * 0.5) * time;
    obj->pos.y += (obj->mov.y + acceleration.y * time * 0.5) * time;

    obj->mov.x += acceleration.x * time;
    obj->mov.y += acceleration.y * time;
    return PHYS_RES_OK;
}

struct phys *phys_create(unsigned int objects_num) {
    const size_t size = sizeof(struct phys) + objects_num * sizeof(struct pobj);
    struct phys *phys = malloc(size);
    if (phys != NULL) {
        memset(phys, 0, size);
        phys->density = 0;
        phys->accel_of_gravity = (struct pvec){.x = 0, .y = 0};
        phys->wind = (struct pvec){.x = 0, .y = 0};
        phys->objects_num = objects_num;
        phys->is_gravity = false;
        phys->time = 0;
    }
    return phys;
}

void phys_destroy(struct phys *phys) {
    free(phys);
}

double phys_get_density(const struct phys *phys) {
    return phys->density;
}

void phys_set_density(struct phys *phys, double density) {
    phys->density = density;
}

struct pvec *phys_ref_accel_of_gravity(struct phys *phys) {
    return &phys->accel_of_gravity;
}

struct pvec *phys_ref_wind(struct phys *phys) {
    return &phys->wind;
}

struct pobj *phys_ref_object(struct phys *phys, unsigned int id) {
    if (id >= phys->objects_num) {
        return NULL;
    }
    return &phys->objects[id];
}

unsigned int phys_get_objects_num(const struct phys *phys) {
    return phys->objects_num;
}

int phys_get_is_gravity(const struct phys *phys) {
    return phys->is_gravity;
}

void phys_set_is_gravity(struct phys *phys, int is_gravity) {
    phys->is_gravity = is_gravity;
}

double phys_get_time(const struct phys *phys) {
    return phys->time;
}

static void compute_object_force(const struct phys *phys, struct pobj *obj) {
    obj->force = (struct pvec){0};

    const struct pvec relative_mov = {
        obj->mov.x - phys->wind.x,
        obj->mov.y - phys->wind.y,
    };

    const double relative_speed = pvec_get_len(&relative_mov);

    if (phys->density != 0 && relative_speed != 0) {
        const double air_f =
            obj->area * phys->density * relative_speed * relative_speed * 0.5 * PHYS_BALL_DRAG_COEF;

        const double k = air_f / relative_speed;
        obj->force.x -= relative_mov.x * k;
        obj->force.y -= relative_mov.y * k;
    }

    const double relative_mass = obj->mass - (obj->volume * phys->density); // maybe negative
    obj->force.x += phys->accel_of_gravity.x * relative_mass;
    obj->force.y += phys->accel_of_gravity.y * relative_mass;
}

static int compute_gravity(struct phys *phys) {
    for (unsigned int i = 0; i < phys->objects_num - 1; i++) {
        for (unsigned int j = i + 1; j < phys->objects_num; j++) {
            struct pobj *a = &phys->objects[i];
            struct pobj *b = &phys->objects[j];

            const struct pvec dist = {
                .x = a->pos.x - b->pos.x,
                .y = a->pos.y - b->pos.y,
            };

            const double dist_len = pvec_get_len(&dist);
            if (dist_len == 0) {
                return PHYS_RES_ERR_ZERO_DIST;
            }

            const double gravy_f = a->mass * b->mass * PHYS_G / (dist_len * dist_len);

            const double k = gravy_f / dist_len;

            a->force.x -= dist.x * k;
            a->force.y -= dist.y * k;
            b->force.x += dist.x * k;
            b->force.y += dist.y * k;
        }
    }
    return PHYS_RES_OK;
}

int phys_run(struct phys *phys, double step_time, unsigned int steps) {
    if (phys->objects_num == 0)
        return PHYS_RES_OK;

    unsigned int counter = 0;
    int err;

    while (counter++ < steps) {
        for (unsigned int i = 0; i < phys->objects_num; i++) {
            compute_object_force(phys, phys->objects + i);
        }
        if (phys->is_gravity && phys->objects_num > 1) {
            err = compute_gravity(phys);
            if (err != PHYS_RES_OK)
                return err;
        }
        for (unsigned int i = 0; i < phys->objects_num; i++) {
            err = pobj_run(phys->objects + i, step_time);
            if (err != PHYS_RES_OK)
                return err;
        }
    }
    phys->time += step_time * steps;
    return PHYS_RES_OK;
}

File: src/PhysGui/FlPhys.cs
using System.Collections;
using System.Runtime.InteropServices;

namespace PhysGui
{
    public static partial class LibFlPhys
    {
        private const string LibraryName = "flphys";

        public const double PHYS_G = 6.6743015151515151514e-11;
        public const double PHYS_PI = 3.1415926535897932385;
        public const double PHYS_AIR_DENSITY = 1.225;
        public const double PHYS_ACCEL_OF_FREE_FALL = 9.80665;
        public const double PHYS_BALL_DRAG_COEF = 0.47;

        public const int PHYS_RES_OK = 0;
        public const int PHYS_RES_ERR_NULL_PTR = -1;
        public const int PHYS_RES_ERR_ZERO_DIST = -2;
        public const int PHYS_RES_ERR_ZERO_MASS = -3;

        [LibraryImport(LibraryName)]
        public static partial int phys_get_is_gravity(IntPtr phys);

        [LibraryImport(LibraryName)]
        public static partial void phys_set_is_gravity(IntPtr phys, int isGravity);

        [LibraryImport(LibraryName)]
        public static partial IntPtr phys_create(int objects_num);

        [LibraryImport(LibraryName)]
        public static partial int phys_run(IntPtr phys, double step_time, long steps);

        [LibraryImport(LibraryName)]
        public static partial int phys_run_bench(IntPtr phys, double step_time, long steps, out double exec_time);

        [LibraryImport(LibraryName)]
        public static partial double phys_get_density(IntPtr phys);

        [LibraryImport(LibraryName)]
        public static partial void phys_set_density(IntPtr phys, double density);

        [LibraryImport(LibraryName)]
        public static partial IntPtr phys_ref_accel_of_gravity(IntPtr phys);

        [LibraryImport(LibraryName)]
        public static partial IntPtr phys_ref_wind(IntPtr phys);

        [LibraryImport(LibraryName)]
        public static partial IntPtr phys_ref_object(IntPtr phys, int id);

        [LibraryImport(LibraryName)]
        public static partial int phys_get_objects_num(IntPtr phys);

        [LibraryImport(LibraryName)]
        public static partial double phys_get_time(IntPtr phys);

        [LibraryImport(LibraryName)]
        public static partial void phys_destroy(IntPtr phys);

        [LibraryImport(LibraryName)]
        public static partial IntPtr pobj_ref_pos(IntPtr pobj);

        [LibraryImport(LibraryName)]
        public static partial IntPtr pobj_ref_mov(IntPtr pobj);

        [LibraryImport(LibraryName)]
        public static partial double pobj_get_mass(IntPtr pobj);

        [LibraryImport(LibraryName)]
        public static partial void pobj_set_mass(IntPtr pobj, double mass);

        [LibraryImport(LibraryName)]
        public static partial double pobj_get_radius(IntPtr pobj);

        [LibraryImport(LibraryName)]
        public static partial void pobj_set_radius(IntPtr pobj, double radius);

        [LibraryImport(LibraryName)]
        public static partial void pvec_set_scs(IntPtr pvec, double len, double angle);

        [LibraryImport(LibraryName)]
        public static partial double pvec_get_x(IntPtr pvec);

        [LibraryImport(LibraryName)]
        public static partial double pvec_get_y(IntPtr pvec);

        [LibraryImport(LibraryName)]
        public static partial void pvec_set_x(IntPtr pvec, double x);

        [LibraryImport(LibraryName)]
        public static partial void pvec_set_y(IntPtr pvec, double y);

        [LibraryImport(LibraryName)]
        public static partial double pvec_get_len(IntPtr pvec);

        [LibraryImport(LibraryName)]
        public static partial double pvec_get_angle(IntPtr pvec);

        [LibraryImport(LibraryName)]
        public static partial IntPtr phys_strerror(int result);
    }


    public sealed class Vector
    {
        private readonly IntPtr _nativePtr;

        internal Vector(IntPtr nativePtr)
        {
            _nativePtr = nativePtr;
        }

        public void SetPolar(double length, double angle) => LibFlPhys.pvec_set_scs(_nativePtr, length, angle);

        public double X
        {
            get => LibFlPhys.pvec_get_x(_nativePtr);
            set => LibFlPhys.pvec_set_x(_nativePtr, value);
        }

        public double Y
        {
            get => LibFlPhys.pvec_get_y(_nativePtr);
            set => LibFlPhys.pvec_set_y(_nativePtr, value);
        }

        public void Set((double x, double y) vector) => (X, Y) = vector;

        public void Deconstruct(out double x, out double y) => (x, y) = (X, Y);

        public static implicit operator (double x, double y)(Vector vector) => (vector.X, vector.Y);

        public override string ToString() => $"({X}, {Y})";

        public double Length => LibFlPhys.pvec_get_len(_nativePtr);
        public double Angle => LibFlPhys.pvec_get_angle(_nativePtr);
    }

    public sealed class PhysicalObject
    {
        private readonly IntPtr _nativePtr;

        internal PhysicalObject(IntPtr nativePtr)
        {
            _nativePtr = nativePtr;
        }

        public Vector Position => new Vector(LibFlPhys.pobj_ref_pos(_nativePtr));
        public Vector Movement => new Vector(LibFlPhys.pobj_ref_mov(_nativePtr));

        public double Mass
        {
            get => LibFlPhys.pobj_get_mass(_nativePtr);
            set => LibFlPhys.pobj_set_mass(_nativePtr, value);
        }

        public double Radius
        {
            get => LibFlPhys.pobj_get_radius(_nativePtr);
            set => LibFlPhys.pobj_set_radius(_nativePtr, value);
        }

    }

    public class PhysException : Exception
    {
        public PhysException(int errorCode)
            : base(GetErrorString(errorCode))
        {
            ErrorCode = errorCode;
        }

        public int ErrorCode { get; }

        private static string GetErrorString(int result)
        {
            var ptr = LibFlPhys.phys_strerror(result);
            return Marshal.PtrToStringUTF8(ptr) ?? $"Unknown error code: {result}";
        }
    }

    public sealed class PhysicsSystem : IDisposable
    {
        private IntPtr _nativePtr;

        public PhysicsSystem(int objects_num)
        {
            _nativePtr = LibFlPhys.phys_create(objects_num);
            if (_nativePtr == IntPtr.Zero)
                throw new InvalidOperationException("Failed to create physics system");
        }

        public double Density
        {
            get => LibFlPhys.phys_get_density(_nativePtr);
            set => LibFlPhys.phys_set_density(_nativePtr, value);
        }

        public Vector AccelerationOfGravity => new Vector(LibFlPhys.phys_ref_accel_of_gravity(_nativePtr));
        public Vector Wind => new Vector(LibFlPhys.phys_ref_wind(_nativePtr));


        private class ObjectList : IReadOnlyList<PhysicalObject>
        {
            private IntPtr _nativePtr;

            public ObjectList(IntPtr nativePtr)
            {
                _nativePtr = nativePtr;
            }

            public PhysicalObject this[int index]
            {
                get
                {
                    if (index < 0 || index >= Count)
                        throw new IndexOutOfRangeException($"Index '{index}' out of range [0, {Count})");
                    return new PhysicalObject(LibFlPhys.phys_ref_object(_nativePtr, index));
                }
            }

            public int Count => LibFlPhys.phys_get_objects_num(_nativePtr);

            public IEnumerator<PhysicalObject> GetEnumerator()
            {
                var count = Count;
                for (int i = 0; i < count; i++)
                {
                    yield return this[i];
                }
            }

            IEnumerator IEnumerable.GetEnumerator()
            {
                return GetEnumerator();
            }
        }

        public IReadOnlyList<PhysicalObject> Objects => new ObjectList(_nativePtr);

        public bool IsGravityEnabled
        {
            get => LibFlPhys.phys_get_is_gravity(_nativePtr) == 0;
            set => LibFlPhys.phys_set_is_gravity(_nativePtr, value ? 1 : 0);
        }

        public double Time => LibFlPhys.phys_get_time(_nativePtr);

        public void Run(double step_time, long steps)
        {
            var res = LibFlPhys.phys_run(_nativePtr, step_time, steps);
            if (res != LibFlPhys.PHYS_RES_OK)
            {
                throw new PhysException(res);
            }
        }

        public double RunBench(double step_time, long steps)
        {
            double execTime;
            var res = LibFlPhys.phys_run_bench(_nativePtr, step_time, steps, out execTime);
            if (res != LibFlPhys.PHYS_RES_OK)
            {
                throw new PhysException(res);
            }
            return execTime;
        }

        public void Dispose()
        {
            if (_nativePtr != IntPtr.Zero)
            {
                LibFlPhys.phys_destroy(_nativePtr);
                _nativePtr = IntPtr.Zero;
            }
        }
    }



}